scripts/ubuntu-bartender/qemu-provider

```diff
@@ -0,0 +1,178 @@
+# Beginning of "public" methods used by ubuntu-bartender
+
+function build-provider-assert-ready {
+  if ! command -v qemu-system-${QEMU_ARCH} &>/dev/null
+  then
+    echo "error: qemu-system-${QEMU_ARCH} was not found in PATH" >&2
+    exit 255
+  fi
+
+  if ! command -v cloud-localds &>/dev/null; then
+    echo "error: cloud-image-utils not found. You can install it with:"
+    echo "sudo apt install -y cloud-image-utils"
+    exit 255
+  fi
+
+  _set_disk_size
+
+  if [ -n "$QEMU_IMAGE" ]; then
+    if [ ! -f "$QEMU_IMAGE" ]; then
+      echo "error: specified image not found at $QEMU_IMAGE"
+      exit 255
+    fi
+    virtual_size=$(qemu-img info --output=json "$QEMU_IMAGE" | jq '."virtual-size"')
+    if [ "$virtual_size" -lt "$QEMU_IMAG_SIZE" ]; then
+      echo "error: qemu disk is too small, it should be at least $(echo $QEMU_IMAG_SIZE | numfmt --to=iec-i) large"
+      echo "Use 'qemu-img resize' to resize your disk"
+      exit 255
+    fi
+  fi
+}
+
+function build-provider-destroy {
+  kill "$(cat "$temp_dir/qemu.pid")" &> /dev/null
+}
+
+function build-provider-create {
+  image="$temp_dir/bionic-server-cloudimg.img"
+  user_data="$temp_dir/user_data.yaml"
+
+  key_dir="$temp_dir/keys"
+  echo "Generating SSH key in $key_dir"
+  mkdir "$key_dir"
+  ssh-keygen -t rsa -f "$temp_dir/keys/ubuntu" -P '' &> /dev/null
+
+  QEMU_EXTRA_OPTIONS="-smp $QEMU_CPU -m $CAP_MEM"
+
+  IMG_ARCH="${QEMU_ARCH}"
+  case "${QEMU_ARCH}" in
+    x86_64)
+      IMG_ARCH='amd64'
+      ;;
+    ppc64le)
+      IMG_ARCH='ppc64el'
+      ;;
+    aarch64)
+      IMG_ARCH='arm64'
+      # there is no default machine for arm64
+      QEMU_EXTRA_OPTIONS="$QEMU_EXTRA_OPTIONS -machine virt -cpu max"
+      # arm only supports uEFI boot
+      QEMU_EXTRA_OPTIONS="$QEMU_EXTRA_OPTIONS -bios /usr/share/qemu-efi-aarch64/QEMU_EFI.fd"
+      ;;
+esac
+
+  if [ "${QEMU_ARCH}" = "$(uname -p)" ] && lsmod | grep -q '^kvm '; then
+    QEMU_EXTRA_OPTIONS="$QEMU_EXTRA_OPTIONS -enable-kvm"
+  fi
+
+  if [ -z "$QEMU_IMAGE" ]; then
+    if ! wget -qO "$image" http://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-$IMG_ARCH.img; then
+      echo "error: could not find the bionic image for $IMG_ARCH on cloud-images.ubuntu.com"
+      cleanup
+      exit 255
+    fi
+
+    qemu-img resize "$image" "$QEMU_IMG_SIZE" &> /dev/null
+  else
+    image="$QEMU_IMAGE"
+  fi
+
+  cat << EOF > "$user_data"
+#cloud-config
+
+ssh_authorized_keys:
+ - $(cat "$temp_dir/keys/ubuntu.pub")
+EOF
+
+  cloud-localds "$temp_dir/seed.img" "$user_data"
+
+  ssh_port=$(_get_ssh_port)
+  qemu-system-${QEMU_ARCH} \
+    $QEMU_EXTRA_OPTIONS \
+    -pidfile "$temp_dir/qemu.pid" \
+    -nographic \
+    -snapshot \
+    -drive if=virtio,format=qcow2,file="${image}" \
+    -nic user,model=virtio,hostfwd="tcp::${ssh_port}-:22" \
+    -drive if=virtio,format=raw,file="$temp_dir/seed.img" &> "$temp_dir/qemu.log" </dev/null &
```

Comment:
there is a `-daemonize` option. with that, you might be able to remove the `sleep` bellow